cabal-install/src/Distribution/Client/ProjectBuilding/UnpackedPackage.hs

```diff
@@ -678,7 +678,31 @@ buildAndInstallUnpackedPackage
           runConfigure
         PBBuildPhase{runBuild} -> do
           noticeProgress ProgressBuilding
-          runBuild
+          -- run preBuildHook. If it returns with 0, we assume the build was
+          -- successful. If not, run the build.
+          code <-
+            rawSystemExitCode
+              verbosity
+              (Just srcdir)
+              "preBuildHook"
+              [ (unUnitId $ installedUnitId rpkg)
+              , (getSymbolicPath srcdir)
+              , (getSymbolicPath builddir)
+              ]
+              `catchIO` (\_ -> return (ExitFailure 10))
+          when (code /= ExitSuccess) $ do
+            runBuild
+          -- not sure, if we want to care about a failed postBuildHook?
```

Comment:
I argue we should. The author of the script can always hide errors of whatever commands they're executing inside the hook.

Reply:
What would you propose?